File: knapsack.cpp
#include <algorithm>
#include <cstddef>
#include <cstdint>
#include <iostream>
#include <ostream>
#include <random>
#include <string>
#include <vector>

#include "genetic_algorithms.hpp"

struct item
{
    item(int64_t value, int64_t weight) : value(value), weight(weight){};

    int64_t value;
    int64_t weight;
};

std::pair<int, int> greedy(std::vector<item> &items, int capacity)
{
    std::sort(items.begin(), items.end(), [](const item &a, const item &b)
              { return (double)a.value / a.weight > (double)b.value / b.weight; });

    int64_t weight = 0;
    int64_t value = 0;
    for (size_t i = 0; i < items.size(); ++i)
    {
        if (weight + items[i].weight <= capacity)
        {
            weight += items[i].weight;
            value += items[i].value;
        }
    }

    return {value, weight};
}

int64_t fitness(const std::vector<int64_t> &genome, const std::vector<item> &items,
                int64_t capacity)
{
    int64_t value = 0;
    int64_t weight = 0;
    for (size_t i = 0; i < genome.size(); ++i)
    {
        value += genome[i] * items[i].value;
        weight += genome[i] * items[i].weight;
    }

    if (weight <= capacity)
        return value;
    else
        return -weight;
}

int main(int argc, const char **argv)
{
    if (argc != 5)
    {
        std::cerr
            << "USAGE: ./knapsack.x <n-items> <population-size> <generations> <mutation-rate>"
            << std::endl;
        return 1;
    }

    // generate random items
    std::random_device rd;
    std::mt19937 engine(rd());
    std::uniform_int_distribution<int64_t> dist(1, 50);

    int64_t total_value = 0;
    int64_t total_weight = 0;
    size_t num_of_items = std::stoul(argv[1]);
    std::vector<item> items;
    for (size_t i = 0; i < num_of_items; ++i)
    {
        items.push_back({dist(engine), dist(engine)});
        total_value += items[i].value;
        total_weight += items[i].weight;
    }

    std::cout << "total value: " << total_value << std::endl;
    std::cout << "total weight: " << total_weight << std::endl;

    int64_t capacity = total_weight * 0.75;
    std::cout << "capacity: " << capacity << std::endl;

    auto solution = greedy(items, capacity);
    std::cout << "************ GREEDY ****************" << std::endl;
    std::cout << "value: " << solution.first << std::endl;
    std::cout << "weight: " << solution.second << std::endl;

    // generate random initial population
    std::cout << "************ GENETIC ****************" << std::endl;
    size_t population_size = std::stoul(argv[2]);
    size_t generations = std::stoul(argv[3]);
    double mutation_rate = std::stod(argv[4]);

    // genetic solver
    ga<int64_t, int64_t> genetic_knapsack(population_size, num_of_items, {0, 1}, generations,
                                          problem_type::maximization);

    genetic_knapsack.generate_population(true);
    genetic_knapsack.evaluate_population(fitness, items, capacity);
    auto best_individual = genetic_knapsack.get_best_individual();
    int64_t value = 0;
    int64_t weight = 0;
    for (size_t i = 0; i < num_of_items; ++i)
    {
        value += best_individual[i] * items[i].value;
        weight += best_individual[i] * items[i].weight;
    }
    std::cout << "value: " << value << std::endl;
    std::cout << "weight: " << weight << std::endl;

    for (size_t g = 0; g < generations; ++g)
    {
        genetic_knapsack.tournament();
        genetic_knapsack.random_crossover();
        genetic_knapsack.random_mutate(mutation_rate);
        genetic_knapsack.evaluate_offsprings(fitness, items, capacity);
        genetic_knapsack.replace();
        genetic_knapsack.evaluate_population(fitness, items, capacity);
    }

    best_individual = genetic_knapsack.get_best_individual();
    value = 0;
    weight = 0;
    for (size_t i = 0; i < num_of_items; ++i)
    {
        value += best_individual[i] * items[i].value;
        weight += best_individual[i] * items[i].weight;
    }
    std::cout << "value: " << value << std::endl;
    std::cout << "weight: " << weight << std::endl;

    return 0;
}

File: test/knapsack.py
import random
import sys

import genetic


class item:
    def __init__(self, value: int, weight: int):
        self.value = value
        self.weight = weight


def greedy(items: list[item], capacity: float):
    # sorting
    items = sorted(items, key=lambda x: x.value / x.weight, reverse=True)

    weight = 0
    value = 0
    for i in items:
        if weight + i.weight <= capacity:
            weight += i.weight
            value += i.value

    return value, weight


def fitness(genome, items, capacity):
    value = 0
    weight = 0
    for i in range(len(genome)):
        value += genome[i] * items[i].value
        weight += genome[i] * items[i].weight

    if weight <= capacity:
        return value
    else:
        return -weight


if __name__ == "__main__":
    if len(sys.argv) != 5:
        print(
            "USAGE: py test.py <n-items> <population-size> <generations> <mutation-rate>"
        )
        exit(-1)

    total_value = 0
    total_weight = 0
    num_of_items = int(sys.argv[1])

    items = []
    for i in range(num_of_items):
        items.append(item(random.randint(1, 100), random.randint(1, 100)))
        total_value += items[i].value
        total_weight += items[i].weight

    capacity = total_weight * 0.75

    value, weight = greedy(items, capacity)
    print("GREEDY")
    print(f"value: {value}")
    print(f"weight: {weight}")

    population_size = int(sys.argv[2])
    generations = int(sys.argv[3])
    mutation_rate = float(sys.argv[4])

    knapsack = genetic.ga(
        population_size,
        num_of_items,
        [0, 1],
        generations,
        genetic.problem_type.maximization,
    )

    knapsack.generate_population()
    knapsack.evaluate_population(fitness, items, capacity)

    best = knapsack.get_best_individual()
    value = 0
    weight = 0
    for i in range(num_of_items):
        value += best.get_genome()[i] * items[i].value
        weight += best.get_genome()[i] * items[i].weight

    print("GENETIC FIRST BEST SOLUTION")
    print(f"value: {value}")
    print(f"weight: {weight}")

    current_best = genetic.individual(best)
    for i in range(generations):
        knapsack.tournament()
        knapsack.random_crossover()
        knapsack.random_mutation(mutation_rate)
        knapsack.evaluate_offsprings(fitness, items, capacity)
        knapsack.replace()
        knapsack.evaluate_population(fitness, items, capacity)
        current_best = knapsack.get_best_individual()
        if current_best > best:
            best = current_best

    value = 0
    weight = 0
    for i in range(num_of_items):
        value += best.get_genome()[i] * items[i].value
        weight += best.get_genome()[i] * items[i].weight

    print("GENETIC BEST SOLUTION")
    print(f"value: {value}")
    print(f"weight: {weight}")
